chalk-ir/src/visit.rs

```diff
@@ -0,0 +1,421 @@
+//! Traits for visiting bits of IR.
+use std::fmt::Debug;
+
+use crate::{
+    BoundVar, DebruijnIndex, DomainGoal, Goal, InferenceVar, Interner, Lifetime, LifetimeData,
+    PlaceholderIndex, ProgramClause, Ty, TyData, WhereClause,
+};
+
+mod binder_impls;
+mod boring_impls;
+pub mod visitors;
+
+pub use visitors::VisitExt;
+
+pub trait VisitResult: Sized {
+    fn new() -> Self;
+
+    fn return_early(&self) -> bool;
```

Comment:
Let's add a comment:

---

Returns true if this result is "complete" and we can stop visiting any further parts of the term. This is used by `FindAny`, for example, to stop the search after a match has been found.